File: java/src/main/java/problems101_200/BinaryTreeLevelOrderTraversal.java
package problems101_200;

import concepts.TreeNode;
import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;
import java.util.Queue;

enum BinaryTreeLevelOrderTraversal {
    DEFAULT {
        @Override
        List<List<Integer>> levelOrder(TreeNode root) {
            List<List<Integer>> lists = new ArrayList<>();
            if (root == null) {
                return lists;
            }

            Queue<TreeNode> queue = new LinkedList<>();
            queue.add(root);
            while (!queue.isEmpty()) {
                int size = queue.size();
                List<Integer> level = new ArrayList<>();
                for (int i = 0; i < size; i++) {
                    TreeNode current = queue.remove();
                    level.add(current.val);
                    if (current.left != null) {
                        queue.add(current.left);
                    }
                    if (current.right != null) {
                        queue.add(current.right);
                    }
                }
                lists.add(level);
            }
            return lists;
        }
    };

    abstract List<List<Integer>> levelOrder(TreeNode root);
}

File: kotlin/src/main/kotlin/problems101_200/BinaryTreeLevelOrderTraversal.kt
package problems101_200

import concepts.TreeNode
import java.util.LinkedList

enum class BinaryTreeLevelOrderTraversal {
    DEFAULT {
        override fun levelOrder(root: TreeNode?): List<List<Int>> {
            val lists = arrayListOf<List<Int>>()
            if (root == null) {
                return lists
            }

            val queue = LinkedList<TreeNode?>()
            queue += root
            while (!queue.isEmpty()) {
                val size = queue.size
                val level = arrayListOf<Int>()
                for (i in 0 until size) {
                    val current = queue.remove()
                    level += current!!.`val`
                    if (current!!.left != null) {
                        queue += current.left
                    }
                    if (current.right != null) {
                        queue += current.right
                    }
                }
                lists += level
            }
            return lists
        }
    },
    ;

    abstract fun levelOrder(root: TreeNode?): List<List<Int>>
}
